### PR TITLE
iterative classical gram-schmidt in eigsh_lanczos

### PR DESCRIPTION
**Pull request by mganahl**
replace the modified gram-schmidt orthogonalization with an iterated classical gram schmidt (faster)
depends on #830, #831

### CODE REVIEW

File: tensornetwork/backends/jax/jax_backend.py [hunk 1]
```diff
@@ -55,10 +55,17 @@ def __init__(self, dtype: Optional[np.dtype] = None) -> None:
     jsp = libjax.scipy
     self.name = "jax"
     self._dtype = np.dtype(dtype) if dtype is not None else None
+    self.jax_precision = precision if precision is not None else libjax.lax.Precision.DEFAULT #pylint: disable=line-too-long
+
+  def configure(self, #pylint: disable=arguments-differ
+                dtype: Optional[np.dtype] = None,
+                precision: Optional[Text] = None):
+    self._dtype = np.dtype(dtype) if dtype is not None else None
+    self.jax_precision = precision if precision is not None else libjax.lax.Precision.DEFAULT #pylint: disable=line-too-long
```
Comment (alewis) [inline tensornetwork/backends/jax/jax_backend.py:64]: you might need to force it to be HIGHEST

File: tensornetwork/backends/jax/jitted_functions_test.py [hunk 2]
```diff
@@ -24,7 +24,8 @@ def matvec(vector, matrix):
   kv = jax.numpy.zeros((ncv + 1, D), dtype=dtype)
   H = jax.numpy.zeros((ncv + 1, ncv), dtype=dtype)
   start = 0
-  kv, H, it, _ = arnoldi(matvec, [mat], x, kv, H, start, ncv, 0.01)
+  precision = jax.lax.Precision.HIGHEST
+  kv, H, it, _ = arnoldi(matvec, [mat], x, kv, H, start, ncv, 0.01, precision)
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions_test.py:28]: The hardcoded 0.01 is a bit weird

Comment (alewis) [reply tensornetwork/backends/jax/jitted_functions_test.py:28]: Or did I write that?

File: tensornetwork/backends/jax/jitted_functions.py [hunk 3]
```diff
@@ -55,83 +62,100 @@ def jax_lanczos(matvec, arguments, init, ncv, neig, landelta, reortho):
...
+          vector.
+        iterations: Number of iterations.
+      Returns:
+        jax.numpy.array: The orthogonalized vector.
+      """
+      vec = vector
+      for _ in range(iterations):
+        ov = jax.numpy.dot(
+            krylov_vectors.conj(), vec, precision=precision)
+        vec = vec - jax.numpy.dot(ov, krylov_vectors)
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:88]: You need the precision in this dot too

File: tensornetwork/backends/jax/jitted_functions.py [hunk 4]
```diff
@@ -55,83 +62,100 @@ def jax_lanczos(matvec, arguments, init, ncv, neig, landelta, reortho):
...
+      beta = jax.numpy.linalg.norm(previous_vector)
+      normalized_vector = previous_vector / beta
+      Av = matvec(jax.lax.reshape(normalized_vector, shape), *arguments)
+      alpha = jax.numpy.vdot(normalized_vector, Av, precision=precision)
+      alphas = alphas.at[i - 1].set(alpha)
+      betas = betas.at[i].set(
+          beta)
+      next_vector = jax.lax.cond(
+          reortho, lambda x: Av, lambda x: jax.numpy.reshape(
+              jax.numpy.ravel(Av) - jax.numpy.ravel(normalized_vector) * alpha -
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:109]: Av.ravel() and so on might make this a bit easier to read. I'd also recommend putting the branch functions on separate lines, but it's a matter of taste.

File: tensornetwork/backends/jax/jitted_functions.py [hunk 5]
```diff
@@ -55,83 +62,100 @@ def jax_lanczos(matvec, arguments, init, ncv, neig, landelta, reortho):
...
+      betas, i = vals[-2], vals[-1]
+      norm = betas[i - 1]
+      return jax.lax.cond(i <= ncv, lambda x: x[0] > x[1], lambda x: False,
+                          [norm, landelta])
+
+    numel = np.prod(shape).astype(np.int32)
+    # note: ncv + 2 because the first vector is all zeros, and the
+    # last is the unnormalized residual.
+    krylov_vecs = jax.numpy.zeros((ncv + 2, numel), dtype=dtype)
+    # NOTE (mganahl): initial state is normalized inside the loop
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:128]: I personally don't like calling the vector a "state" in these functions, but again it's a matter of taste.

File: tensornetwork/backends/jax/jitted_functions.py [hunk 6]
```diff
@@ -276,6 +278,28 @@ def _arnoldi_fact(matvec, args, v0, krylov_vectors, H, start, num_krylov_vecs,
...
+          `n`: integer denoting the column-position of the overlap
+            <`krylov_vector`|`vector`> within `H`.
+      Returns:
+        updated vals.
+  
+      """
+      vector, krylov_vectors, n, H = vals
+      v = krylov_vectors[j, :]
+      h = jax.numpy.vdot(v, vector, precision=precision)
+      H = jax.ops.index_update(H, jax.ops.index[j, n], h)
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:299]:
```suggestion
      H = H.at[j, n].set(h)
```

**Review by alewis — COMMENTED** [review]
Looks good, few small suggestions

File: tensornetwork/backends/jax/jitted_functions.py (hunk 3 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:88]: good catch, thanks!

**Comment by mganahl** [comment]
sorry, just noticed that this was branched off the wrong branch, need to close it!

File: tensornetwork/backends/jax/jax_backend.py (hunk 1 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jax_backend.py:64]: I'd like to leave it at whatever the default value is in JAX

File: tensornetwork/backends/jax/jitted_functions_test.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions_test.py]: That was me, its a threshold value. Added a variable name

**Comment by mganahl** [comment]
@alewis ping for review

**Comment by mganahl** [comment]
@alewis ping for review, thanks!

File: tensornetwork/backends/jax/jitted_functions.py [hunk 7]
```diff
@@ -43,111 +43,133 @@ def _generate_jitted_eigsh_lanczos(jax: types.ModuleType) -> Callable:
...
+    Lanczos iteration for symmeric eigenvalue problems. If reortho = False,
+    the Krylov basis is constructed without explicit re-orthogonalization. 
+    In infinite precision, all Krylov vectors would be orthogonal. Due to 
+    finite precision arithmetic, orthogonality is usually quickly lost. 
+    For reortho=True, the Krylov basis is explicitly reorthogonalized.
     Args:
       matvec: A callable implementing the matrix-vector product of a
         linear operator.
       arguments: Arguments to `matvec` additional to an input vector.
         `matvec` will be called as `matvec(init, *args)`.
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:57]: Something weird is happening with the Markdown here

File: tensornetwork/backends/jax/jitted_functions.py [hunk 8]
```diff
@@ -43,111 +43,133 @@ def _generate_jitted_eigsh_lanczos(jax: types.ModuleType) -> Callable:
...
-      v = krylov_vectors[i, :]
-      vector -= jax.numpy.vdot(
-          v, vector, precision=precision) * jax.numpy.reshape(v, vector.shape)
-      return [vector, krylov_vectors]
+    def iterative_classical_gram_schmidt(
+        vector: jax.ShapedArray,
+        krylov_vectors: jax.ShapedArray,
+        iterations: int = 2) -> jax.ShapedArray:
+      """
+      orthogonalize `vector`  to all rows of `krylov_vectors`.
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions.py:78]: here too

File: tensornetwork/backends/jax/jitted_functions_test.py [hunk 9]
```diff
@@ -24,8 +24,8 @@ def matvec(vector, matrix):
   kv = jax.numpy.zeros((ncv + 1, D), dtype=dtype)
   H = jax.numpy.zeros((ncv + 1, ncv), dtype=dtype)
   start = 0
-
-  kv, H, it, _ = arnoldi(matvec, [mat], x, kv, H, start, ncv, 0.01, precision)
+  tol = 0.01
```
Comment (alewis) [inline tensornetwork/backends/jax/jitted_functions_test.py:27]: I didn't notice in my last comment this was a test, sorry

**Review by alewis — APPROVED** [review]

File: tensornetwork/backends/jax/jitted_functions.py (hunk 7 above)
Comment (mganahl) [reply tensornetwork/backends/jax/jitted_functions.py:57]: There are a couple of places where this happends. I'll have to run through all files eventually to fix this

**Pull request merged**